src/tests/java/htsjdk/samtools/SAMRecordUnitTest.java

```diff
@@ -90,8 +91,283 @@ public void testOffsetAtReference(String cigar, int posInReference, int expected
+        sam = new SAMRecordSetBuilder().addFrag("test", 0, 1, false, false, "3S9M", null, 2);
+        sam.setMateReferenceName("unresolvableName");
+        Assert.assertTrue(sam.getMateReferenceName().equals("unresolvableName"));
+        Assert.assertTrue(sam.getMateReferenceIndex().equals(SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX));
+    }
+
+    @Test
+    public void testMateReferenceIndex() {
+        // NO_ALIGNMENT_REFERENCE
+        SAMRecord sam = new SAMRecordSetBuilder().addFrag("test", 0, 1, false, false, "3S9M", null, 2);
+        sam.setMateReferenceIndex(SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX);
+        Assert.assertTrue(sam.getMateReferenceIndex().equals(SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX));
+        Assert.assertTrue(sam.getMateReferenceName().equals(SAMRecord.NO_ALIGNMENT_REFERENCE_NAME));
+
+        // valid reference
+        sam = new SAMRecordSetBuilder().addFrag("test", 0, 1, false, false, "3S9M", null, 2);
+        sam.setMateReferenceIndex(3);
+        Assert.assertTrue(sam.getMateReferenceIndex().equals(3));
+        Assert.assertTrue(sam.getMateReferenceName().equals("chr4"));
+    }
+
+    @Test(expectedExceptions=IllegalArgumentException.class)
+    public void testInvalidMateReferenceIndex() {
+        // unresolvable reference
+        SAMRecord sam = new SAMRecordSetBuilder().addFrag("test", 0, 1, false, false, "3S9M", null, 2);
+        sam.setMateReferenceIndex(9999);
+    }
+
+    // ----------------- NULL header tests ---------------------
+
+    @Test
+    public void testNullHeaderReferenceName() {
+        SAMRecord sam = new SAMRecordSetBuilder().addFrag("test", 0, 1, false, false, "3S9M", null, 2);
+        final SAMFileHeader samHeader = sam.getHeader();
+        Assert.assertTrue(null != samHeader);
+        final String originalRefName = sam.getReferenceName();
+        int originalRefIndex = sam.getReferenceIndex();
+
+        sam.setHeader(null);
+        // setting header to null retains the previously assigned ref name
+        Assert.assertTrue(originalRefName.equals(sam.getReferenceName()));
+        // resetting the header to non-null re-resolves the reference index back to the original
+        sam.setHeader(samHeader);
+        Assert.assertTrue(sam.getReferenceIndex().equals(originalRefIndex));
+
+        // null header allows reference name to be set to NO_ALIGNMENT_REFERENCE_NAME
+        sam.setHeader(null);
+        sam.setReferenceName(SAMRecord.NO_ALIGNMENT_REFERENCE_NAME);
+        Assert.assertTrue(sam.getReferenceName().equals(SAMRecord.NO_ALIGNMENT_REFERENCE_NAME));
+        Assert.assertTrue(sam.getReferenceIndex().equals(SAMRecord.NO_ALIGNMENT_REFERENCE_INDEX));
+
+        // null header allows any ref name set
+        final String unresolvableRefName = "unresolvable";
+        sam.setReferenceName(unresolvableRefName);
+        Assert.assertTrue(unresolvableRefName.equals(sam.getReferenceName()));
+    }
+
+    @Test
+    public void testNullHeaderReferenceIndex() {
```

Comment:
`testNullHeaderReferenceName` and `testNullHeaderReferenceIndex` are not well decomposed -- they both appear to be testing reference name and index-related functionality in equal amounts. Recommend either merging them or better separating the testing of the reference name and index-related methods.